// typed-vuex/lib/index.es.js
const getStoreType = (store) => {
  return {
    actionContext: {},
    rootState: {},
    rootGetters: {},
    storeInstance: {}
  };
};

const actionTree = (_store, tree) => tree;

const getterTree = (_state, tree) => tree;

const mutationTree = (_state, tree) => tree;

const getAccessorType = (store2) => {
  return {};
};
const getNestedState = (parent, namespaces) => {
  if (!parent[namespaces[0]]) {
    return parent;
  } else {
    return getNestedState(parent[namespaces[0]], namespaces.slice(1));
  }
};
const createAccessor = (store2, {
  getters,
  state: state2,
  mutations,
  actions,
  namespaced
}, namespace = "") => {
  const namespacedPath = namespace && namespaced ? `${namespace}/` : "";
  const accessor = {};
  Object.keys(getters || {}).forEach((getter) => {
    Object.defineProperty(accessor, getter, {
      get: () => store2.getters[`${namespacedPath}${getter}`]
    });
  });
  const evaluatedState = state2 ? typeof state2 === "function" ? state2() : state2 : {};
  Object.keys(evaluatedState).forEach((prop) => {
    if (!Object.getOwnPropertyNames(accessor).includes(prop)) {
      const namespaces = namespace.split("/");
      Object.defineProperty(accessor, prop, {
        get: () => getNestedState(store2.state, namespaces)[prop]
      });
    }
  });
  Object.keys(mutations || {}).forEach((mutation) => {
    accessor[mutation] = (mutationPayload) => store2.commit(`${namespacedPath}${mutation}`, mutationPayload);
  });
  Object.keys(actions || {}).forEach((action) => {
    accessor[action] = (actionPayload) => store2.dispatch(`${namespacedPath}${action}`, actionPayload);
  });
  return accessor;
};
const useAccessor = (store2, input, namespace) => {
  const accessor = createAccessor(store2, input, namespace);
  Object.keys(input.modules || {}).forEach((moduleNamespace) => {
    const nestedNamespace = namespace ? `${namespace}/${moduleNamespace}` : moduleNamespace;
    accessor[moduleNamespace] = useAccessor(store2, input.modules[moduleNamespace], nestedNamespace);
  });
  return accessor;
};
const getAccessorFromStore = (pattern) => {
  return (store2) => useAccessor(store2, pattern._modules.root._rawModule);
};

export { actionTree, getAccessorFromStore, getAccessorType, getStoreType, getterTree, mutationTree, useAccessor };
